### PR TITLE
Added a custom dialog in the QGIS data source manager

### LINKED ISSUE #13: hook into QGIS data source manager dialog
Add our initial search GUI, as implemented in #2 to the QGIS data source manager dialog, as researched in #4.

The main objective of this issue is to have our custom GUI show up when a user has the plugin loaded and opens up the QGIS data source manager. No need to actually connect our GUI's buttons to any actions at this point

### PR DESCRIPTION
**Pull request by Samweli**
fixes #13

### CODE REVIEW

File: src/qgis_geonode/gui/geonode_provider.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+from qgis.core import QgsProject
+from qgis.gui import QgsSourceSelectProvider, QgsAbstractDataSourceWidget
+
+from qgis.PyQt.uic import loadUiType
+
+WidgetUi, _ = loadUiType(os.path.join(os.path.dirname(__file__), '../ui/qgis_geonode_main_ui.ui'))
+
+
+class GeonodeProvider(QgsSourceSelectProvider):
```
Comment (ricardogsilva) [inline src/qgis_geonode/gui/geonode_provider.py:32]: how about you name this `GeonodeSourceSelectProvider` instead? Seems more inline with QGIS naming

File: src/qgis_geonode/qgisgeonode/main.py [hunk 2]
```diff
@@ -174,6 +182,8 @@ def initGui(self):
             callback=self.run,
             parent=self.iface.mainWindow())
 
+        QgsGui.sourceSelectProviderRegistry().addProvider(self.geonodeProvider)
+
```
Comment (ricardogsilva) [inline src/qgis_geonode/qgisgeonode/main.py:186]: From my local tests this only seems to add/remove the provider to the QGIS data source manager dialog after restarting QGIS.

I mean that when the plugin is installed (or manually enabled/disabled in the QGIS plugin manager) there is no change to the QGIS data source manager dialog. Only after restarting QGIS do I see a change.

Perhaps there is a signal that needs to be emitted? Looking at the [QGIS API docs](https://qgis.org/api/classQgsSourceSelectProviderRegistry.html), it seems it is not possible to dynamically reload the list of providers. Did you do any research on this?

What are our options here?

**Comment by Samweli** [comment]
> Also, please don't just add stuff remove also unused code. I am thinking especially of the `main.py` module, that seems to be accumulating a lot of unused stuff related to the dock widget, and also the `Qgis_GeoNode_dockwidget.py` file too, which seems to be irrelevant now.

Yes, can I do this on a separate PR?

File: src/qgis_geonode/qgisgeonode/main.py (hunk 2 above)
Comment (Samweli) [reply src/qgis_geonode/qgisgeonode/main.py:186]: > From my local tests this only seems to add/remove the provider to the QGIS data source manager dialog after restarting QGIS

@ricardogsilva I confirm this

> Did you do any research on this?

No I didn't 

> What are our options here?

I'm taking a look at the QGIS API, will let you know once I find out what we need to do.

Comment (Samweli) [reply src/qgis_geonode/qgisgeonode/main.py]: @ricardogsilva I have looked at the implementation details of "addProvider()" and "removeProvider()" seems there is no code that work on updating dynamically the provider list on the data source manager dialog other than updating (by adding or removing) the provider list.

We could make effort to add signal to the https://qgis.org/api/classQgsSourceSelectProviderRegistry.html as you have suggested above and connect to it in https://qgis.org/api/classQgsDataSourceManagerDialog.html in order to update the dialog providers list.

**Review by ricardogsilva — CHANGES_REQUESTED** [review]
Almost ready, just a couple of small changes left:

- Please clean up the code in `main.py` and remove unused gui files - I prefer having that in this PR, since it relates to the same thing
- Please remove the `src/__init__.py` file as it should not be needed (I'm guessing probably your IDE added that by default?)

As for the load/reload of the sourceselect provider, since this is something that needs changing in QGIS core, lets leave it like this for now. We will contribute the change to QGIS core first and then eventually adjust our code here to  make use of it.

**Pull request merged**